File: apps/patter-voice-sdk/docs/examples/conversation-history.ts
/**
 * Conversation history — agent remembers the entire call.
 *
 * on_transcript fires after every turn with the latest text and full history.
 * on_call_end fires when the call finishes with the complete transcript.
 */
import { Patter, Twilio, OpenAIRealtime } from "getpatter";

interface TranscriptEntry {
  role: "user" | "assistant";
  text: string;
}

interface TranscriptEvent {
  role: "user" | "assistant";
  text: string;
  history: TranscriptEntry[];
}

interface CallEndEvent {
  transcript: TranscriptEntry[];
  duration_seconds?: number;
}

async function main() {
  const phone = new Patter({
    carrier: new Twilio(),                              // TWILIO_* from env
    phoneNumber: "+15550001234",
    webhookUrl: "xxx.ngrok-free.dev",
  });

  const agent = phone.agent({
    engine: new OpenAIRealtime({ voice: "alloy" }),     // OPENAI_API_KEY from env
    systemPrompt:
      "You are a helpful assistant. " +
      "Reference earlier parts of the conversation when relevant.",
    firstMessage:
      "Hello! Let's have a conversation. I'll remember everything we discuss.",
  });

  await phone.serve({
    agent,
    port: 8000,

    // Called after every turn with latest text and full history
    onTranscript: async (data: TranscriptEvent) => {
      console.log(`[${data.role}] ${data.text}`);
      console.log(`  History length: ${data.history.length} turns`);
    },

    // Called when the call finishes — full transcript available here
    onCallEnd: async (data: CallEndEvent) => {
      const transcript = data.transcript ?? [];
      console.log("\n=== Call Summary ===");
      for (const entry of transcript) {
        console.log(`  ${entry.role}: ${entry.text}`);
      }
      console.log(`  Total turns: ${transcript.length}`);
    },
  });
}

main().catch(console.error);

File: apps/patter-voice-sdk/docs/examples/conversation-history.py
"""Conversation history — agent remembers the entire call."""
import asyncio
from getpatter import Patter, Twilio, OpenAIRealtime


async def on_transcript(data: dict) -> None:
    """Called after every turn with the latest text and full history."""
    role = data["role"]
    text = data["text"]
    history = data.get("history", [])
    print(f"[{role}] {text}")
    print(f"  History length: {len(history)} turns")


async def on_call_end(data: dict) -> None:
    """Called when the call finishes. Full transcript is available here."""
    transcript = data.get("transcript", [])
    print("\n=== Call Summary ===")
    for entry in transcript:
        print(f"  {entry['role']}: {entry['text']}")
    print(f"  Total turns: {len(transcript)}")


async def main():
    phone = Patter(
        carrier=Twilio(),                               # TWILIO_* from env
        phone_number="+15550001234",
        webhook_url="xxx.ngrok-free.dev",
    )

    agent = phone.agent(
        engine=OpenAIRealtime(voice="alloy"),           # OPENAI_API_KEY from env
        system_prompt="You are a helpful assistant. "
                      "Reference earlier parts of the conversation when relevant.",
        first_message="Hello! Let's have a conversation. I'll remember everything we discuss.",
    )

    await phone.serve(
        agent=agent,
        port=8000,
        on_transcript=on_transcript,
        on_call_end=on_call_end,
    )


asyncio.run(main())
